File: AndroidClient/app/src/main/java/com/mancersoft/litevpn/transport/IVpnTransport.java
package com.mancersoft.litevpn.transport;

import java.io.Closeable;
import java.util.concurrent.CompletableFuture;

public interface IVpnTransport {

    interface MessageListener {
        void onMessage(Packet packet);
    }

    interface ClosedListener {
        void onClosed(boolean isByUser);
    }

    Closeable createSocket() throws Exception;

    boolean isReliable();

    CompletableFuture<Boolean> connect();

    void disconnect();

    void sendAsync(Packet packet);

    void setOnMessageListener(MessageListener messageListener);

    void setOnClosedListener(ClosedListener closedListener);
}

File: AndroidClient/app/src/main/kotlin/com/mancersoft/litevpn/transport/IVpnTransport.kt
package com.mancersoft.litevpn.transport

import java.io.Closeable
import java.util.concurrent.CompletableFuture

interface IVpnTransport {

    @Throws(Exception::class)
    fun createSocket(): Closeable

    val isReliable: Boolean

    fun connect(): CompletableFuture<Boolean>

    fun disconnect()

    fun sendAsync(packet: Packet)

    fun setOnMessageListener(messageListener: ((Packet) -> Unit)?)

    fun setOnClosedListener(closedListener: ((isByUser: Boolean) -> Unit)?)
}